### PR TITLE
Java: Add ZADD and ZINCR commands. (Sorted Set Commands)

### PR DESCRIPTION
**Pull request by SanHalacogluImproving**

### CODE REVIEW

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java [hunk 1]
```diff
@@ -9,14 +9,28 @@
...
+    /**
+     * Converts a map of string keys and values of any type into an array of strings with alternating
+     * values and keys.
+     *
+     * @param args Map of string keys to values of any type to convert.
+     * @return Array of strings [value1.toString(), key1, value2.toString(), key2, ...].
+     */
+    public static String[] convertMapToValueKeyStringArray(Map<String, ?> args) {
+        return args.entrySet().stream()
+                .flatMap(entry -> Stream.of(entry.getValue().toString(), entry.getKey()))
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/utils/ArrayTransformUtils.java:33]: You can introduce a second argument to swap fields or nor like
```suggestion
    public static String[] convertMapToStringArray(Map<String, ?> args, boolean swap) {
        return args.entrySet().stream()
                .flatMap(entry -> swap ? Stream.of(entry.getValue().toString(), entry.getKey()
                : Stream.of(entry.getKey(), entry.getValue().toString())))
```

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java [hunk 2]
```diff
@@ -34,4 +48,16 @@ public static <T, U extends T> U[] castArray(T[] objectArr, Class<U> clazz) {
...
+    /**
+     * Concatenates multiple arrays of type T and returns a single concatenated array.
+     *
+     * @param arrays Varargs parameter for arrays to be concatenated.
+     * @param <T> The type of the elements in the arrays.
+     * @return A concatenated array of type T.
+     */
+    @SafeVarargs
+    public static <T> T[] concatenateArrays(T[]... arrays) {
+        return Stream.of(arrays).flatMap(Stream::of).toArray(size -> Arrays.copyOf(arrays[0], size));
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/utils/ArrayTransformUtils.java:61]: Can you try this?
```suggestion
        return Stream.of(arrays).flatMap(Stream::of).toArray(T[]::new);
```

File: java/client/src/main/java/glide/api/models/commands/ZaddOptions.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    private final ConditionalChange conditionalChange;
+
+    /**
+     * Specifies conditions for updating scores with zadd {@link SortedSetBaseCommands#zadd} command.
+     */
+    private final UpdateOptions updateOptions;
+
+    @RequiredArgsConstructor
+    @Getter
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:33]: Do you use these getters? If not, probably you are missing a test. If you still don't use them -delete.

File: java/client/src/main/java/glide/api/models/commands/ZaddOptions.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+     * Converts ZaddOptions into a String[].
+     *
+     * @return String[]
+     */
+    public String[] toArgs() {
+        if (conditionalChange == ConditionalChange.ONLY_IF_DOES_NOT_EXIST && updateOptions != null) {
+            throw new IllegalArgumentException(
+                    "The GT, LT, and NX options are mutually exclusive. Cannot choose both "
+                            + updateOptions.redisApi
+                            + " and NX.");
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:76]: You need tests which cover all 4 combinations. You can update client tests by using different combinations there + add UT for this class (for coverage).

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java (hunk 1 above)
Comment (jonathanl-bq) [reply java/client/src/main/java/glide/utils/ArrayTransformUtils.java:33]: I'm actually not a fan of adding boolean flags for configuring like this usually. It's adding another branch instruction and it doesn't really save much in terms of code reuse.

File: java/client/src/main/java/glide/api/models/commands/ZaddOptions.java (hunk 3 above)
Comment (jonathanl-bq) [reply java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:33]: Good point. We can probably delete them.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 5]
```diff
@@ -13,6 +13,7 @@ public class TransactionTestUtilities {
     private static final String key3 = "{key}" + UUID.randomUUID();
     private static final String key4 = "{key}" + UUID.randomUUID();
     private static final String key5 = "{key}" + UUID.randomUUID();
+    private static final String key6 = "{key}" + UUID.randomUUID();
```
Comment (jonathanl-bq) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:16]: I'm not a huge fan of this numbered key convention we've developed. It's annoying to reconcile merge conflicts when the key numbers overlap when two different people add new tests here.

**Review by jonathanl-bq — APPROVED** [review]
Approved, but please address Yury's comments at least. My comment on variable naming conventions in the TransactionTestUtilities can be discussed further with the rest of the team, but not a blocker for this PR.

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java (hunk 2 above)
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/utils/ArrayTransformUtils.java:61]: It does not allow for generic array creation this way.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 5 above)
Comment (SanHalacogluImproving) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:16]: Following what they have on Py and Node tests.

File: java/client/src/main/java/glide/api/models/commands/ZaddOptions.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:73]: Add a UT which covers this case. It is good to create another UT class for zaddoptions only.

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/main/java/glide/api/models/commands/ZaddOptions.java (hunk 4 above)
Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:73]: I did add a test for it in RedisClientTest

Comment (SanHalacogluImproving) [reply java/client/src/main/java/glide/api/models/commands/ZaddOptions.java:73]: https://github.com/Bit-Quill/glide-for-redis/pull/115/files#diff-976ae060a9ceecb773b36ba8fa33376c9bb012f2697e105aceda55a25b191b4bR914

**Pull request merged**